Add support for simplecov's new expected JSON output

fede-moya opened the pull request:
Following https://github.com/codeclimate/test-reporter/issues/413, this PR attempts to add support to simplecov `0.20.0` +. As discussed in the issue shared before, the main idea is that simplecov will natively incorporate a JSON formatter, which we will be triggered by default when running simplecov within an environment that has the `CC_TEST_REPORTER_ID` variable present. 

Accordingly, the simplecov formatter will now try to read coverage information from the output of the new simplecov's JSON formatter. This new report will contain more information reporting as "ignored" those lines marked with `:nocov:`.

The current implementation pretends to maintain support for previous versions, therefore there are two different paths for looking up the json report, the old `.resultset.json` and the new `coverage.json`, a priority in order has been given to `coverage.json`. When a path is given, like when using `test-reporter format-coverate -t simplecov -p /some-path` the formatter will try to read the indicated information as if was outputted by the new JSON formatter, in case of that not being the case it will fallback to reading as if it has the same format as `.resultset.json`. The idea is that no one gets affected by this changes. 

As a final note,  Until https://github.com/simplecov-ruby/simplecov/pull/923 get's merged or a variation of it, It wouldn't make much sense to merge this work, though in theory it wouldn't cause problems to anyone.

### Addresses:
- https://github.com/codeclimate/test-reporter/issues/246
- https://github.com/codeclimate/test-reporter/issues/413
- (_potentially_) https://github.com/codeclimate/test-reporter/issues/385

### Bonus track
- This pull request includes a minimal change on the circleci configuration, was necessary to get CI passing. It just bumps up the XCODE version of the macos engines.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/codeclimate/test-reporter?pullRequest=441) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/codeclimate/test-reporter?pullRequest=441) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/codeclimate/test-reporter?pullRequest=441) it.</sub>

noelia-lencina commented on formatters/simplecov/simplecov.go:
```diff
@@ -41,30 +43,71 @@ func (r Formatter) Format() (formatters.Report, error) {
+	if err != nil {
+		return legacyFormat(r, rep)
+	} else {
+		logrus.Debugf("Analyzing simplecov json output from latest format %s", r.Path)
```
nit: would it be possible to move this logic into its own file similar to `legacyFormat` to keep the same structure?

fede-moya replied:
I liked the suggestion thanks, addressed. Just couldn't find a more clever name to the new formatter than `jsonFormatter` although it makes sense since it designed to read ouput from the the SimpleCov's JsonFormatter.

noelia-lencina reviewed (approved):
Nice work! Tests look great 👏 I left a really small comment, feel free to ignore it if you feel it's not worth it.

fede-moya commented:
@filipesperandio can I get your review here please ? I'm posting here since I cannot add you as a reviewer, I think it's due to the repo's permissions settings

filipesperandio commented:
@fede-moya: You and @noelia-lencina should have write access to the repo now. 🤦

filipesperandio commented on formatters/simplecov/simplecov.go:
```diff
@@ -1,17 +1,17 @@
-var searchPaths = []string{"coverage/.resultset.json"}
+var simpleCovJsonFormatterPath, legacyPath string = "coverage/coverage.json", "coverage/.resultset.json"
+
+var searchPaths = []string{simpleCovJsonFormatterPath, legacyPath}
```
Can this be inlined? Doesn't look like we use those vars anywhere else.

fede-moya replied:
Done 👍🏼

filipesperandio commented on formatters/simplecov/simplecov.go:
```diff
@@ -32,39 +32,16 @@ func (f *Formatter) Search(paths ...string) (string, error) {
-	if err != nil {
-		return rep, errors.WithStack(errors.Errorf("could not open coverage file %s", r.Path))
-	}
+	rep, err = jsonFormat(r,rep)
```
I'm confused. Does this override what line 34 does?

fede-moya replied:
Do you mind taking a new look ? I don't think I'm following the the question sorry

filipesperandio replied:
I might have something confused in my head, not seeing a problem here anymore 🤦

filipesperandio commented on formatters/simplecov/legacy_formatter.go:
```diff
@@ -0,0 +1,47 @@
+	Coverage map[string]formatters.Coverage `json:"coverage"`
+}
+
+func legacyFormat(r Formatter, rep formatters.Report) (formatters.Report, error) {
```
Seems like we always return `rep`, can we avoid that and only return `error`?
And the caller deal with the return it needs to pass? Like `return rep, legacyFormat(r, rep)`?

filipesperandio replied:
Might not be a `go` thing, though... let me know.

fede-moya replied:
I liked your suggestion so I tested it out, but for some unknown reason to me tests just started to fail without a clear explanation. Check https://github.com/codeclimate/test-reporter/pull/441/commits/dacd033784cd68e6b9b23275ccd6d7f2efc3233e , then I reverted the commit and tests passed again. So ...

filipesperandio commented on formatters/simplecov/simplecov_test.go:
```diff
@@ -9,6 +9,33 @@ import (
+	rep, err := f.Format()
+	r.NoError(err)
+
+	r.Len(rep.SourceFiles, 7)
```
The single char vars here are making my eye itchy  🤨 
Can we make the names a bit more descriptive? Like, 4 lines down and I don't recall what `r` is anymore 😬

fede-moya replied:
hahah, yeah I know sorry, I kinda follow the pattern that was already in place, but I have just update all the tests to use better namings.

fede-moya commented:
Hey @filipesperandio just addressed your comments, thanks for your, as usual,  detailed review. I'm just pinging you again in this comment to make some noise on your notifications so it became the first thing you see on Monday and luckily if you have further comments we can address them tomorrow 💪🏼

filipesperandio commented:
Worked. 😂

filipesperandio commented on formatters/simplecov/simplecov_test.go:
```diff
@@ -18,21 +18,21 @@ func Test_ParseLegacy(t *testing.T) {
 	}
 
-	r := require.New(t)
+	assert := require.New(t)
```
Nice! thanks!

filipesperandio commented on .circleci/config.yml:
```diff
@@ -24,7 +24,7 @@ jobs:
   test_macos:
     macos:
-      xcode: "9.0"
+      xcode: "11.3.0"
```
Any chance we can do this in a separate PR? Feel free to tag me.
Just want to isolate infra/tooling changes from other things, makes it easier to revert in case we need to.

fede-moya replied:
yeap, I think that makes sense. Will do that now
